### ISSUE TITLE
[JENKINS-41960] Developer can change stage and parallel level configuration

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#3781]
<div class="panel" style="border-width: 1px;"><div class="panelHeader" style="border-bottom-width: 1px;background-color: #3878de;color: white;"><b>Improvement on roadmap</b></div><div class="panelContent">
<p>This improvement is on the Blue Ocean project roadmap. Check the <a href="https://jenkins.io/projects/blueocean/roadmap/" class="external-link" target="_blank" rel="nofollow noopener">roadmap page</a> for updates. </p>
</div></div>

<p><b>Scope</b><br/>
<em>Stage configuration</em></p>
<ul>
	<li>Clicking anywhere on the stage name
	<ul>
		<li>Developer can change parallel name</li>
		<li>Developer can change settings for
		<ul>
			<li>Agent</li>
			<li>Environment</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

<p><em>Parallel configuration</em></p>
<ul>
	<li>Clicking on a parallel shows the config for the parallel
	<ul>
		<li>Developer can change parallel name</li>
		<li>Developer can change settings for
		<ul>
			<li>Agent</li>
			<li>Environment</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

<p><em>New style for environment</em></p>
<ul>
	<li>There is a new style for environment<br/>
<span class="image-wrap" style=""><a id="38130_thumb" href="https://issues.jenkins.io/secure/attachment/38130/38130_env.png" title="env.png" file-preview-type="image" file-preview-id="38130" file-preview-title="env.png"><img src="https://issues.jenkins.io/secure/thumbnail/38130/_thumb_38130.png" style="border: 0px solid black" role="presentation"/></a></span> </li>
</ul>

<p><b>Design</b><br/>
<em>Pipeline Settings</em></p>
<ul>
	<li><a href="https://invis.io/Q2A110KM9#/233893917_Editor-Home_W_Parallels" class="external-link" target="_blank" rel="nofollow noopener">https://invis.io/Q2A110KM9#/233893917_Editor-Home_W_Parallels</a></li>
	<li><a href="https://zpl.io/1h9jFU" class="external-link" target="_blank" rel="nofollow noopener">https://zpl.io/1h9jFU</a></li>
</ul>

<p><em>Stage</em></p>
<ul>
	<li><a href="https://invis.io/Q2A110KM9#/233884743_Editor-Reorder_Parallels" class="external-link" target="_blank" rel="nofollow noopener">https://invis.io/Q2A110KM9#/233884743_Editor-Reorder_Parallels</a></li>
	<li><a href="https://zpl.io/Z1FptFb" class="external-link" target="_blank" rel="nofollow noopener">https://zpl.io/Z1FptFb</a></li>
</ul>

<p><em>Animation & timing</em></p>
<ul>
	<li><a href="http://codepen.io/brody/pen/gWZvrr" class="external-link" target="_blank" rel="nofollow noopener">http://codepen.io/brody/pen/gWZvrr</a></li>
</ul>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay">jamesdumay</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41960" target="_blank">Developer can change stage and parallel level configuration</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kzantow.png" title="kzantow's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow">kzantow</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: blueocean-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-09-25T00:32:44+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 5
<li><b>imported</b>: 2025-12-01
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<div class="panel" style="border-width: 1px;"><div class="panelHeader" style="border-bottom-width: 1px;background-color: #3878de;color: white;"><b>Improvement on roadmap</b></div><div class="panelContent">
<p>This improvement is on the Blue Ocean project roadmap. Check the <a href="https://jenkins.io/projects/blueocean/roadmap/" class="external-link" target="_blank" rel="nofollow noopener">roadmap page</a> for updates. </p>
</div></div>

<p><b>Scope</b>
<em>Stage configuration</em></p>
<ul>
	<li>Clicking anywhere on the stage name
	<ul>
		<li>Developer can change parallel name</li>
		<li>Developer can change settings for
		<ul>
			<li>Agent</li>
			<li>Environment</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

<p><em>Parallel configuration</em></p>
<ul>
	<li>Clicking on a parallel shows the config for the parallel
	<ul>
		<li>Developer can change parallel name</li>
		<li>Developer can change settings for
		<ul>
			<li>Agent</li>
			<li>Environment</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

<p><em>New style for environment</em></p>
<ul>
	<li>There is a new style for environment
<span class="image-wrap" style=""><a id="38130_thumb" href="https://issues.jenkins.io/secure/attachment/38130/38130_env.png" title="env.png" file-preview-type="image" file-preview-id="38130" file-preview-title="env.png"><img src="https://issues.jenkins.io/secure/thumbnail/38130/_thumb_38130.png" style="border: 0px solid black" role="presentation"/></a></span> </li>
</ul>

<p><b>Design</b>
<em>Pipeline Settings</em></p>
<ul>
	<li><a href="https://invis.io/Q2A110KM9#/233893917_Editor-Home_W_Parallels" class="external-link" target="_blank" rel="nofollow noopener">https://invis.io/Q2A110KM9#/233893917_Editor-Home_W_Parallels</a></li>
	<li><a href="https://zpl.io/1h9jFU" class="external-link" target="_blank" rel="nofollow noopener">https://zpl.io/1h9jFU</a></li>
</ul>

<p><em>Stage</em></p>
<ul>
	<li><a href="https://invis.io/Q2A110KM9#/233884743_Editor-Reorder_Parallels" class="external-link" target="_blank" rel="nofollow noopener">https://invis.io/Q2A110KM9#/233884743_Editor-Reorder_Parallels</a></li>
	<li><a href="https://zpl.io/Z1FptFb" class="external-link" target="_blank" rel="nofollow noopener">https://zpl.io/Z1FptFb</a></li>
</ul>

<p><em>Animation &amp; timing</em></p>
<ul>
	<li><a href="http://codepen.io/brody/pen/gWZvrr" class="external-link" target="_blank" rel="nofollow noopener">http://codepen.io/brody/pen/gWZvrr</a></li>
</ul>
</pre>
</details>
</details>
<details><summary><i>2 attachments</i></summary>

- [env.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38130/env.png)
  > ![env.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38130/env.png)
- [stage_level_config.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/35906/stage_level_config.png)
  > ![stage_level_config.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/35906/stage_level_config.png)
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote links associated with this issue:_


  - [PR 1375](https://github.com/jenkinsci/blueocean-plugin/pull/1375)
  - [Page](https://wiki.jenkins.io/pages/viewpage.action?pageId=98403475)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41960?focusedId=292774&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-292774">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow" class="user-hover" rel="kzantow">kzantow</a> if we aren't aiming for this for 1.0, I think this could be an early editor thing for <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> to take a look at at some point (would help get familiar with internals) - if it makes sense with you. </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow" class="user-hover" rel="kzantow">kzantow</a> if we aren't aiming for this for 1.0, I think this could be an early editor thing for <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> to take a look at at some point (would help get familiar with internals) - if it makes sense with you. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kzantow.png" title="kzantow's avatar" /> kzantow:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41960?focusedId=310076&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310076">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>FYI - this isn't blocking <a href="https://issues.jenkins.io/browse/JENKINS-44390" title="Support for editing parallel stages" class="issue-link" data-issue-key="JENKINS-44390"><del>JENKINS-44390</del></a>, if anything the other way around.</p>

<p>Changes for this ticket should be very simple, I originally implemented it but then had to put some code in place to hide this from the parallel nodes because it wasn't supported, this should be trivial to apply the same stage configuration panel.</p>

<p>Adding the remaining configuration items e.g. <em>while</em> will be the only thing that would require a bit of time</p>
</pre>
</details>

</ul>
</details>

<p>FYI - this isn't blocking <a href="https://issues.jenkins.io/browse/JENKINS-44390" title="Support for editing parallel stages" class="issue-link" data-issue-key="JENKINS-44390"><del>JENKINS-44390</del></a>, if anything the other way around.</p>

<p>Changes for this ticket should be very simple, I originally implemented it but then had to put some code in place to hide this from the parallel nodes because it wasn't supported, this should be trivial to apply the same stage configuration panel.</p>

<p>Adding the remaining configuration items e.g. <em>while</em> will be the only thing that would require a bit of time</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41960?focusedId=310413&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310413">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>BTW <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a>  I think this one maybe can't be done until the linked blocking ticket is done - ie editor supports the "new" style of parallel stages. Without that, there is no parallel config that can be stored. With it, parallel config is just stage config, but on each parallel stage. So perhaps sync up with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow" class="user-hover" rel="kzantow">kzantow</a> before starting on this when the time comes. </p>

<p>Maybe: <a href="https://issues.jenkins-ci.org/browse/JENKINS-44390" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-44390</a> is a duplicate of this? I am confused. </p>
</pre>
</details>

</ul>
</details>

<p>BTW <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a>  I think this one maybe can't be done until the linked blocking ticket is done - ie editor supports the "new" style of parallel stages. Without that, there is no parallel config that can be stored. With it, parallel config is just stage config, but on each parallel stage. So perhaps sync up with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow" class="user-hover" rel="kzantow">kzantow</a> before starting on this when the time comes. </p>

<p>Maybe: <a href="https://issues.jenkins-ci.org/browse/JENKINS-44390" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-44390</a> is a duplicate of this? I am confused. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/cliffmeyers.png" title="cliffmeyers's avatar" /> cliffmeyers:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41960?focusedId=310822&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310822">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Assigning to Keith as it looks like from discussion that this ticket is a partial dup or overlaps with <a href="https://issues.jenkins.io/browse/JENKINS-44390" title="Support for editing parallel stages" class="issue-link" data-issue-key="JENKINS-44390"><del>JENKINS-44390</del></a>. If these aren't something you can take on soon just let me know and we can bounce them back my way.</p>
</pre>
</details>

</ul>
</details>

<p>Assigning to Keith as it looks like from discussion that this ticket is a partial dup or overlaps with <a href="https://issues.jenkins.io/browse/JENKINS-44390" title="Support for editing parallel stages" class="issue-link" data-issue-key="JENKINS-44390"><del>JENKINS-44390</del></a>. If these aren't something you can take on soon just let me know and we can bounce them back my way.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kzantow.png" title="kzantow's avatar" /> kzantow:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41960?focusedId=311862&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-311862">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Not exactly a duplicate, but will be handled by <a href="https://issues.jenkins.io/browse/JENKINS-44390" title="Support for editing parallel stages" class="issue-link" data-issue-key="JENKINS-44390"><del>JENKINS-44390</del></a></p>
</pre>
</details>

</ul>
</details>

<p>Not exactly a duplicate, but will be handled by <a href="https://issues.jenkins.io/browse/JENKINS-44390" title="Support for editing parallel stages" class="issue-link" data-issue-key="JENKINS-44390"><del>JENKINS-44390</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kzantow.png" title="kzantow's avatar" /> kzantow:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41960?focusedId=312834&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-312834">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Implemented this separately to isolate the change(s) just a bit. Went by what I could scrape up regarding mockups on this ticket.</p>
</pre>
</details>

</ul>
</details>

<p>Implemented this separately to isolate the change(s) just a bit. Went by what I could scrape up regarding mockups on this ticket.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-43172%22">JENKINS-43172</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-44390%22">JENKINS-44390</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-36291%22">JENKINS-36291</a>]</i>
